Fix usability of clear all chips button

Kiarokh opened the pull request:
fix #936 

## Review:
- [ ] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [ ] Commits have the correct *type* for the changes made
- [ ] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [ ] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

adrianschmidt commented on src/components/chip-set/chip-set.tsx:
```diff
@@ -572,9 +578,11 @@ export class ChipSet {
                 tabindex="0"
                 role="button"
+                title={this.clearAllLabel}
+                area-label={this.clearAllLabel}
```
```suggestion
                aria-label={this.clearAllLabel}
```

Kiarokh replied:
typical me :D

adrianschmidt commented on src/style/mixins.scss:
```diff
@@ -67,3 +67,26 @@
+    }
+
+    &:hover {
+        background-color: #757580;
```
While we're at it, shouldn't we use the new color-palette for these background colours? 🙂

Kiarokh replied:
I think we will go through these color things later, one by one in each component.

adrianschmidt commented on src/components/chip-set/chip-set.tsx:
```diff
@@ -82,6 +82,12 @@ export class ChipSet {
     public searchLabel: string;
 
+    /**
+     * Clear all label to display as `title` and `area-label` for icon button that clears all chips
```
```suggestion
     * Label to display as `title` and `aria-label` for button that clears all chips
```

adrianschmidt commented on src/components/chip-set/chip-set.tsx:
```diff
@@ -572,9 +578,11 @@ export class ChipSet {
         return (
             <i
```
I think this should be a link. Then there should be no need for the `tabindex="0"`, and it ought to be possible to trigger it by tabbing to it and pressing \<enter\>.

Alternatively, we might want to use a limel-icon-button, but I'm guessing that might be a little too bloated for use here?

Kiarokh replied:
have you tried to see whether having it as a link would work? Otherwise I'lll test tomorrow ;)

Kiarokh replied:
@adrianschmidt this didn't work unfortunately. :( maybe for a link to work, we need a URL? donno

Kiarokh commented:
@adrianschmidt the tabbing behavior seems to be inconsistent. Sometimes it works (specially when the field is "focused" and you're adding or removing chips) But some times it doesn't (like when you tab into the field from outside). We couldn't really figure out how and why.

Kiarokh commented:
> It's because unless you hover the component, the button is set to `display:none` as soon as the input loses focus.

But we don't want the (x) to show up when the field is not being used (is not focused). However, in some situations as soon as you tab, the field will lose focus. Right now I'm not sure how to fix this problem. Perhaps the easiest way is to constantly show the (x) button as long as `.has-chips` class exists, even if the field is not in focus.

adrianschmidt commented:
> > It's because unless you hover the component, the button is set to `display:none` as soon as the input loses focus.
> 
> But we don't want the (x) to show up when the field is not being used (is not focused). However, in some situations as soon as you tab, the field will lose focus. Right now I'm not sure how to fix this problem. Perhaps the easiest way is to constantly show the (x) button as long as `.has-chips` class exists, even if the field is not in focus.

Invisible and `display:none` are not the same thing. But I can fix it later. Busy with some other things at the moment, but tell me once you've done what you want and can, and I'll look at the remaining bits.

Kiarokh commented:
> By "invisible", I didn't mean `visibility:hidden` or any other specific styling, just that there are ways to make something be not possible to see, while still possible to focus. 

No problem Adrian. Using `visibility:hidden` was actually a smart idea, even if you didn't mean it 😄. `https://developer.mozilla.org/en-US/docs/Web/CSS/visibility` they say:
![image](https://user-images.githubusercontent.com/35954987/92112190-b51f9780-eded-11ea-9825-f220c248aeff.png)

Anyhow, I later realized that you can still tab into the (x) even if there are no chips (due to lack of CSS rules), so in the [last fixup commit](https://github.com/Lundalogik/lime-elements/pull/938/commits/7ee6cb3005a85457ce1b3b0bec540ee8373af194), I fixed this. I also found a bug which is hard to explain. The styles caused the (x) to jump up and down so you couldn't really click it. This happened when the first line in the chipset was filled with chips and the `<input>` (where you type in) was placed on the next line. 

Anyhow, it's now fixed by having an absolut position for the (x) which only considers `right` and `top`. (doesn't middle align itself based on the height of the chipset). This fix gives a more pleasing UI as well.

jgroth commented on src/components/chip-set/chip-set.tsx:
```diff
@@ -82,6 +82,12 @@ export class ChipSet {
+     * Label to display as `title` and `aria-label` for button that clears all chips
+     */
+    @Prop({ reflectToAttr: true })
+    public clearAllLabel: string;
```
Is this to make it translatable or do we want consumers to be able to change the label to something completely different? 

If it's only for translation purposes, I would suggest to remove this for now and just have it in English. We need to discuss how to handle translations in Lime Elements at a broader scope before it becomes unmaintainable. I would prefer if we could solve it without having props for all translation strings on a component level. Thoughts? :slightly_smiling_face:

adrianschmidt replied:
> Is this to make it translatable or do we want consumers to be able to change the label to something completely different?

Ping @Kiarokh ☝️ 

> If it's only for translation purposes, I would suggest to remove this for now and just have it in English. We need to discuss how to handle translations in Lime Elements at a broader scope before it becomes unmaintainable. I would prefer if we could solve it without having props for all translation strings on a component level. Thoughts? 🙂

It seems we already have some temporary hardcoded translations in Lime Elements, used for the date-picker. If it's for translation, I would suggest adding the translations there for now.

I absolutely agree that we should discuss how to handle translations generally 👍

Kiarokh replied:
> Ping @Kiarokh ☝️ 

thanks

> Is this to make it translatable

yes :) 

> handle translations in Lime Elements at a broader scope

would be awesome, decent, amazing, professional ;) 

> solve it without having props for all translation strings on a component level

you guys know the best how to do it optimally

adrianschmidt replied:
Great! I'll add some hardcoded translations 👍

adrianschmidt replied:
Ah, crap… no, that's just going to require us to send in the language code instead. I'm going with English only for now.

Unfortunately, since that property has already been merged and released, this is now becoming a breaking change 😭

adrianschmidt replied:
I decided to deprecate the property, and give it a default value in English. That way we can avoid making a breaking change for now.

adrianschmidt replied:
Sorry for being confused. The property was introduced in this PR, so I just removed it and used a hardcoded value in English instead.

jgroth commented on src/components/chip-set/chip-set.tsx:
```diff
@@ -83,10 +83,10 @@ export class ChipSet {
 
     /**
-     * Label to display as `title` and `aria-label` for button that clears all chips
+     * ***DEPRECATED*** Label to display as `title` and `aria-label` for button that clears all chips
```
```suggestion
     * Label to display as `title` and `aria-label` for button that clears all chips
     * @deprecated
```

adrianschmidt replied:
Resolving this as I removed the property.

jgroth commented:
I'm not sure why GitHub is not letting me respond in the thread so I'm forced to write my reply here instead...

> Unfortunately, since that property has already been merged and released, this is now becoming a breaking change :sob:

Has it been merged? :thinking:  It's part of this PR is it not? I'm confused now :smile:

maltja commented:
It doesn't seem to work for me to focus on the clear all chips button with tab in Firefox and Safari. Chrome works perfect.

adrianschmidt commented:
> I'm not sure why GitHub is not letting me respond in the thread so I'm forced to write my reply here instead...
> 
> > Unfortunately, since that property has already been merged and released, this is now becoming a breaking change 😭
> 
> Has it been merged? 🤔 It's part of this PR is it not? I'm confused now 😄

Oh, yeah, it is! I thought it had just been renamed in this one, but introduced in the PR where the button was added. I'll just remove the prop then! 🎉

adrianschmidt commented:
> It doesn't seem to work for me to focus on the clear all chips button with tab in Firefox and Safari. Chrome works perfect.

Good catch! I'll take a look at it 👍

adrianschmidt commented:
> > It doesn't seem to work for me to focus on the clear all chips button with tab in Firefox and Safari. Chrome works perfect.
> 
> Good catch! I'll take a look at it 👍

I could get it to work in Firefox if I changed it from an `<a>` to a `<button>`, but then I lost the hover effect. Safari refused to focus it whatever I did, but I think Safari generally doesn't focus on any links or buttons?

I've now spent most of the day on this without really getting anywhere, and I think we should just merge this as it's clearly better than what's currently on master. If we wish to spend more time polishing it later, we can do so in a separate PR.

Kiarokh commented:
I also think we should merge it. We will at some point go through keyboard navigation and focus styles of all components. These things we can certainly improve later.

However, it might be a good idea to already create an issue regarding Safari in this board https://github.com/orgs/Lundalogik/projects/61 so we don't forget them.

adrianschmidt commented:
It actually works in Safari, but you have to enable this accessibility option in Safari's preferences:

![image](https://user-images.githubusercontent.com/154725/92699980-8bff8980-f34e-11ea-9574-8d7d1f061e48.png)
